File: algorithms/102. Binary Tree Level Order Traversal/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        def dfs(root, h):
            if root:
                if h == len(result):
                    result.append([])
                result[h].append(root.val)
                dfs(root.left, h + 1)
                dfs(root.right, h + 1)
        result = []
        dfs(root, 0)
        return result

File: algorithms/102. Binary Tree Level Order Traversal/solution.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> result;
        dfs(root, 0, result);
        return result;
    }

    void dfs(TreeNode * root, int height, vector<vector<int>> & result) {
        if ( root ) {
            if ( height == result.size() ) {
                result.push_back(vector<int>());
            }
            result[height].push_back(root->val);
            dfs(root->left, height + 1, result);
            dfs(root->right, height + 1, result);
        }
    }
};
